### PR TITLE
#4347 handle the display of remote vaccinations when no internet available

### LINKED ISSUE #4347: Handle the display of remote vaccinations when no internet available
## Is your feature request related to a problem? Please describe.

Title sounds a bit nonsensical but bear with me..

As part of the work for the Kiribati COVID-19 rollout we did a couple of things: 
- allowed the download of all global patients and their medical history to each tablet's local storage. 
- added a patient medical history button to the patient selection list in the vaccination module.

The idea is that a patient's medical history is available to view even if there is no internet connection (as it has been stored locally) .

This new patient medical history button correctly displays the medical history. 

However, there are also 3 other places where patient medical history can be accessed, and **these 3 other places do not display the patient's medical history correctly when there is no internet available**. 

The 3 other places are: 
- The Dispensing module patient medical history button
- The Dispensing module vaccination lookup in the ADR form
- The patient medical history button in the final step of the Vaccination module. 

For each of the above places local vaccinations are displayed correctly but remote vaccinations (i.e. vaccinations administered to the patient in a different store) do not appear if there is no internet available. The lookup appears to be attempting to retrieve from the server rather than looking in the local storage. 

See screenshots: 

1. New patient medical history button in Vaccine module

2. Internet off. Local and remote vaccinations correctly displayed: 

image

3. Dispensing module patient medical history button: 

image

4. Internet off. Only local vaccinations show. 

image

5. Dispensing module ADR vaccination lookup: 

image

6. Internet off. Only local vaccinations show: 

image

7. Vaccination module final step patient medical history button: 

image

8. Internet off. Only local vaccinations show: 

image

## Describe the solution you'd like

The 3 other places to display a patient's medical history in a similar fashion to the new Vaccine module patient medical history (i.e. to retrieve remote vaccinations from local storage if available). 

## Implementation

**A description of how the solution would best be implemented in the code base**

## Describe alternatives you've considered

**A clear and concise description of any alternative solutions or features you've considered.**

## Additional context

Screenshots taken from 8.2.2

### PR DESCRIPTION
**Pull request by katherine-sussol**
Fixes #4347 

Will break Tonga's vaccine history until #4384 is completed (and run on server before upgrade).

## Change summary
- Removed vaccine history from the end of the vaccine dispensing user flow
- Removed vaccine-specific history modal - merged name note logic into PatientHistory modal
- Update vaccine history to merge local name notes with remotely fetched vaccinations (if mSupply pref to sync patients everywhere is set, remote lookup shouldn't be done for vaccines)
- Update dispensing history to merge local name notes and local transactions for history
- Off-road to fix vaccine dispensing button from not showing when no modules enabled 
- Small off-road to remote fetching, if no results were found, the local results were being overwritten with an empty array

## Testing

Steps to reproduce or otherwise test the changes of this PR:

Needs PatientEvent with code 'vaccination' set up in mSupply Desktop
- [ ] ADR vaccine history, vaccine dispensing history and regular dispensing history should all display both local and remote vaccinations if pref to view all patients history is set 
- [ ] If 'Patients sync everywhere' is enabled the remote vaccinations should show up offline (assuming version of mSupply >= Kiribati version)
- [ ] Dispensing history should also still fetch both local and remote regular dispensing history records.
- [ ] There should be no duplicates when doing a remote fetch

### Related areas to think about
- Not sure if I broke anything dispensing related - did notice medicine administrators and prescribers seem to be 'N/A' a lot from the API during testing (but this is mobile so didn't look into it too far...)

### CODE REVIEW

File: src/database/utilities/preferenceConstants.js [hunk 1]
```diff
@@ -16,6 +16,7 @@ export const PREFERENCE_KEYS = {
   HIDE_SNAPSHOT_COLUMN: 'usesHideSnapshotColumn',
   CAN_EDIT_PATIENTS_FROM_ANY_STORE: 'can_edit_patients_from_any_store',
   CAN_VIEW_ALL_PATIENTS_HISTORY: 'can_view_all_patients_history',
+  NEW_PATIENTS_VISIBLE_THIS_STORE_ONLY: 'new_patients_visible_in_this_store_only',
```
Comment (katherine-sussol) [inline src/database/utilities/preferenceConstants.js:19]: Based this on https://github.com/sussol/msupply/issues/9254#issuecomment-888164810 - I'm not too sure if this is safe enough to use 🤔

File: src/selectors/Entities/name.js [hunk 2]
```diff
@@ -77,16 +79,47 @@ export const selectCanEditPatient = state => {
...
+  const { id: vaccinationPatientEventID } = vaccinationPatientEvent ?? {};
+
+  const nameNotes = patient?.nameNotes
+    ?.filter(
+      ({ patientEventID, data }) =>
+        patientEventID === vaccinationPatientEventID && validateJsonSchemaData(jsonSchema, data)
+    )
+    .map(({ data: vaccinationNameNotes }) => ({
+      ...vaccinationNameNotes,
+      doses: 1, // Currently not possible to dispense more than 1 dose
```
Comment (katherine-sussol) [inline src/selectors/Entities/name.js:117]: There were no quantities in the name_notes created 😅 (maybe we should add one?)

File: src/hooks/useLocalAndRemoteHistory.js [hunk 3]
```diff
@@ -23,11 +23,27 @@ const reducer = (state, action) => {
       const { payload } = action;
       const { data } = payload;
       const { data: initialData } = state;
-      const localIds = initialData.map(history => history.id);
+      const localTransactionIds = initialData.map(transactions => transactions.id);
+
+      // Name notes do not have transaction ids, create filter instead based on itemCode/confirmDate
```
Comment (katherine-sussol) [inline src/hooks/useLocalAndRemoteHistory.js:28]: This has a big assumption that someone can't receive two doses of the same vaccine in the same day to do duplicate detection between local/remote transactions 🤔  - not sure if there is a better way to filter.

**Review by katherine-sussol — COMMENTED** [review]
Just also realised quantities are a bit weird now in the history (because vaccination name_notes don't have quantities set):
image

Maybe this should show doses for vaccines rather than totalQuantity?

File: src/selectors/Entities/name.js (hunk 2 above)
Comment (joshxg) [reply src/selectors/Entities/name.js:117]: Hehe, yeah, we should. For now it's fine to assume `1` 😁  Technically the whole prescription is in there !!! 🤣

File: src/hooks/useLocalAndRemoteHistory.js (hunk 3 above)
Comment (joshxg) [reply src/hooks/useLocalAndRemoteHistory.js:28]: there is `extra.prescription.id` and we can add the id to the name note similar to doses.

So `id = nameNote.id || extra.prescription.id` ? Bit yuck but

Comment (joshxg) [reply src/hooks/useLocalAndRemoteHistory.js:28]: oops more like.. `id = nameNote.transactionId || nameNote.extra.prescription.id` - hopefully you get the idea .. 😛

Comment (katherine-sussol) [reply src/hooks/useLocalAndRemoteHistory.js:28]: Ooh, I see what you mean, the whole prescription is supposed to be in the nameNote 😮  that would make life a lot easier - I think at the moment it's not (or at least not on my environment unless I've done something weird)...

Prescription gets created first, and then theoretically the name note should take the prescription as an arg:
https://github.com/openmsupply/mobile/blob/75d8f47cdfc5ce78786b29f4e5d8fc326a36152d/src/actions/Entities/VaccinePrescriptionActions.js#L273-L289

Except it looks like `createPrescription` doesn't actually return the prescription?
https://github.com/openmsupply/mobile/blob/75d8f47cdfc5ce78786b29f4e5d8fc326a36152d/src/actions/Entities/VaccinePrescriptionActions.js#L145-L171

**Review by joshxg — APPROVED** [review]
cooooooool

File: src/pages/MenuPage.js [hunk 4]
```diff
@@ -260,7 +265,7 @@ const Menu = ({
 
   return (
     <View style={{ ...appBackground }}>
-      {usingModules ? <ModuleLayout /> : <OriginalLayout />}
+      {usingModules || hasVaccines ? <ModuleLayout /> : <OriginalLayout />}
```
[the remark this answers is not in the document]
Comment (katherine-sussol) [reply src/pages/MenuPage.js]: Yeahh, that's way better, thanks! f24d2b3

**Comment by katherine-sussol** [comment]
Have also updated the dispensing history view to use the doses as the quantity for vaccines (hopefully not too confusing? I think it's less confusing than decimal quantities - and also the decimal quantities are currently not available via name notes...) 800b646

image

Will create a separate issue to save the prescription data

**Pull request merged**